File: Daily_Challenge/20221127_LC_2225.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.27

class Solution {
public:
    vector<vector<int>> findWinners(vector<vector<int>>& matches) {
        vector<vector<int>> ans(2);
        unordered_set<int> seen;
        unordered_map<int, vector<int>> count;
        int max_value = 0;
        for (int i = 0; i < matches.size(); i++){
            int winner = matches[i][0], loser = matches[i][1];
            count[loser].push_back(winner);
            seen.insert(winner);
            seen.insert(loser);
            max_value = max(max_value, max(winner, loser));
        }
        for (int i = 1; i <= max_value; i++){
            if (!seen.count(i)){
                continue;
            }
            if (!count.count(i)){
                ans[0].push_back(i);
            }
            else if (count[i].size() == 1){
                ans[1].push_back(i);
            }
        }
        return ans;
    }
};

File: Daily_Challenge/20221127_LC_2225.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.27

class Solution:
    def findWinners(self, matches: List[List[int]]) -> List[List[int]]:
        ans = [[] for _ in range(2)]
        count = collections.defaultdict(list)
        seen = set()
        max_value = 0
        for winner, loser in matches:
            count[loser].append(winner)
            seen.add(winner)
            seen.add(loser)
            max_value = max(max_value, winner, loser)
        
        for item in range(1, max_value+1):
            if item not in seen:
                continue
            if item not in count:
                ans[0].append(item)
            elif len(count[item]) == 1:
                ans[1].append(item)
        return ans